algebra/euclidean_domain.lean

```diff
@@ -0,0 +1,295 @@
+/- 
+Copyright (c) 2018 Louis Carlin. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Louis Carlin
+
+Euclidean domains and Euclidean algorithm (extended to come)
+A lot is based on pre-existing code in mathlib for natural number gcds
+-/
+import tactic.ring
+
+universe u
+
+class euclidean_domain (α : Type u) [decidable_eq α] extends integral_domain α :=
+(quotient : α → α → α)
+(remainder : α → α → α)
+(quotient_mul_add_remainder_eq : ∀ a b, (quotient a b) * b + (remainder a b) = a) -- This could be changed to the same order as int.mod_add_div. We normally write qb+r rather than r + qb though.
+(valuation : α → ℕ)
+(valuation_remainder_lt : ∀ a b, b ≠ 0 → valuation (remainder a b) <  valuation b)
+(le_valuation_mul : ∀ a b, b ≠ 0 → valuation a ≤ valuation (a*b))
+
+/-
+le_valuation_mul is often not a required in definitions of a euclidean domain since given the other properties we can show there is a (noncomputable) euclidean domain α with the property le_valuation_mul.
+So potentially this definition could be split into two different ones (euclidean_domain_weak and euclidean_domain_strong) with a noncomputable function from weak to strong.
+I've currently divided the lemmas into strong and weak depending on whether they require le_valuation_mul or not.
+-/
+
+namespace euclidean_domain
+variable {α : Type u}
+variables [decidable_eq α] [euclidean_domain α]
+
+instance : has_div α := ⟨quotient⟩
+
+instance : has_mod α := ⟨remainder⟩
+
+instance : has_sizeof α := ⟨valuation⟩
+
+lemma gcd_decreasing (a b : α) (w : a ≠ 0) : has_well_founded.r (b % a) a := valuation_remainder_lt b a w
+
+def gcd : α → α → α
+| a b := if a_zero : a = 0 then b
+     else have h : has_well_founded.r (b % a) a := gcd_decreasing a b a_zero,
+        gcd (b%a) a
+
+/- weak lemmas -/
+
+@[simp] lemma mod_zero (a : α) : a % 0 = a :=
+begin
+  have := euclidean_domain.quotient_mul_add_remainder_eq a 0,
+  simp at this,
+  exact this,
+end
+
+lemma dvd_mod_self (a : α) : a ∣ a % a :=
+begin
+  let d := (a/a)*a, -- ring tactic can't solve things without this
+  have : a%a = a - (a/a)*a, from
+    calc
+      a%a = d + a%a  - d : by ring
+      ... = (a/a)*a + a%a - (a/a)*a : by dsimp [d]; refl
+      ... = a - (a/a)*a : by simp [(%), (/), quotient_mul_add_remainder_eq a a],
+  rw this,
+  exact dvd_sub (dvd_refl a) (dvd_mul_left a (a/a)),
+end
+
+lemma mod_lt  : ∀ (a : α) {b : α}, valuation b > valuation (0:α) →  valuation (a%b) < valuation b :=
+begin
+  intros a b h,
+  by_cases b_zero : (b=0),
+  {
+    rw b_zero at h,
+    have := lt_irrefl (valuation (0:α)),
+    contradiction,
+  },
+  {
+    exact valuation_remainder_lt a b b_zero,
+  }
+end
+
```

Comment:
I'm not suggesting it's any better than what you have, but just to show you what this looks like in term mode (I used simpa again instead of the rw) I just did the translation and got

```
lemma mod_lt  : ∀ (a : α) {b : α}, valuation b > valuation (0:α) →  valuation (a%b) < valuation b :=
λ a b h, dite (b=0) 
  (λ b_zero, false.elim $ lt_irrefl (valuation (0:α)) (by simpa [b_zero] using h))
  (valuation_remainder_lt a b)
```

I know that this is the sort of proof that the mathlib people actually seem to write in practice (note: I am not one of the mathlib people!) but my point is simply that if, like I was, you are addicted to tactic mode, then I should encourage you to start thinking about what features of tactic mode you are actually using (probably the fact that the goal and hypotheses at each stage are clearly written in the messages box is the main one! It certainly was for me) and then consider the puzzle of unravelling a short tactic proof back into term mode. Funnily enough this is the precise opposite of the advice I give to students just starting out, where I can't get them moving into tactic mode quickly enough so they can see what is going on without being bewildered by the whole functional side of things.